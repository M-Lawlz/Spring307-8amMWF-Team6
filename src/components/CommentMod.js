import React from 'react';
import TextField from '@material-ui/core/TextField';
import App from "firebase/app";

export default class CommentMod extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            newComment : "",
            user: null,
            userData: null,
            tourId : this.props.tourId,
            currentTourComments : null
        }
    }

    async componentDidMount() {
        await App.auth().onAuthStateChanged((user) => {
            if (user) {
              this.setState({ user: user });
              this.fetchUserInfo();
            }
          });
        const firebase = require("firebase");
        const db = firebase.firestore();
        const toursDb = db.collection("Tours");
    
        toursDb.get().then(snapshot => {
            snapshot.forEach(doc => {
                if(doc.data().tourId == this.state.tourId) {
                    this.setState({
                        currentTourComments : doc.data().comments
                    });
                }
            });
        });
    }

    fetchUserInfo = () => {
        const userDoc = App.firestore()
          .collection("users")
          .doc(this.state.user.email);
        userDoc
          .get()
          .then((doc) => {
            if (doc.exists) {
              this.setState({ userData: doc.data() });
            }
          })
          .catch((error) => {
            console.error("Error: ", error.message);
          });
      };

    handleSubmit = (event) => {
        const firebase = require("firebase");
        const db = firebase.firestore();
        const currentTour = db.collection("Tours").doc(this.state.tourId.toString());
        const timeRn = firebase.firestore.Timestamp.now().toDate();

        /* TODO: Figure out timestamp printing issue, keeps thinking its 1972 */
        const newItem = {comment : this.state.newComment,
            userName : this.state.userData.username,
            commentDate : timeRn}
        currentTour.update({
            comments : firebase.firestore.FieldValue.arrayUnion(newItem)
        });
        var updatedComments = this.state.currentTourComments;
        updatedComments.push(newItem);
        this.setState({
            currentTourComments : updatedComments,
            newComment: ""
        });
        event.preventDefault();
    }

    updateComment = (event) => {
        this.setState({
            newComment: event.target.value
        });
    }

    render() {
        return (
            <div className="container">
                <h1>Comments!</h1>
                <hr></hr>
                <br/>
                <div className="commentBox">
                    <h2>Join the Discussion!</h2>
                    <form className="commentForm" onSubmit={this.handleSubmit}>
                        <TextField className="commentText" type="text"
                        rows="2" onChange={this.updateComment}
                        value={this.state.newComment}
                        placeholder="Drop a comment!" disabled={this.state.user === null}
                        required>
                        </TextField>
                        {this.state.user === null ? 
                        <button type="submit" disabled>Login to Comment!</button> :
                        <button type="submit">Post Comment!</button>}
                    </form>
                    <div>
                        {
                            (this.state.currentTourComments !== null && this.state.currentTourComments !== undefined) ?
                            <div>
                                {this.state.currentTourComments.map(com => {
                                    return (
                                        <div>
                                        <hr></hr>
                                            <span>
                                            <div className="commentSection">
                                                <b>
                                                    {com.userName + ": "}
                                                </b>
                                                {com.comment + " "}
                                            </div>
                                            </span>
                                        </div>
                                    )
                                })}
                            </div>
                            : null
                        }
                    </div>
                </div>
            </div>
        )
    }
}