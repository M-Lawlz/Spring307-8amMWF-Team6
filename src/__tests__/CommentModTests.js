import React from "react";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import CommentMod from "../components/CommentMod";

test("CommentMod renders successfully", () => {
  const commentMod = render(<CommentMod location={{ pathname: "Test" }} />);
  expect(toJson(commentMod)).toMatchSnapshot();
});

test("Updating a comment", () => {
  const comment = shallow(<CommentMod location={{ pathname: "Test" }} />);
  const instance = comment.instance();
  instance.updateComment({ target: { value : "newcomment"}});
  expect(instance.state.newComment).toBe("newcomment");
});

test("Updating the tour page's comment", () => {
  const comment = shallow(<CommentMod location={{ pathname: "Test" }} />);
  const instance = comment.instance();
  instance.updateLocalTourComments([{comment: "comment1", userName: "johncena", commentDate: "01/01/2000"}]);
  expect(instance.state.currentTourComments).toStrictEqual([{comment: "comment1", userName: "johncena", commentDate: "01/01/2000"}]);
});