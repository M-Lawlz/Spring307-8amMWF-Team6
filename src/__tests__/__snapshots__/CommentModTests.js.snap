// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentMod renders successfully 1`] = `
<div
  className="container"
>
  <h1>
    Comments!
  </h1>
  <hr />
  <br />
  <div
    className="commentBox"
  >
    <h2>
      Join the Discussion!
    </h2>
    <form
      className="commentForm"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        className="commentText"
        disabled={true}
        onChange={[Function]}
        placeholder="Drop a comment!"
        required={true}
        rows="2"
        type="text"
        value=""
      />
      <button
        disabled={true}
        type="submit"
      >
        Login to Comment!
      </button>
    </form>
    <div />
  </div>
</div>
`;
